script/state_manager.js

```diff
@@ -0,0 +1,440 @@
+				$SM.remove('perks');
+			}
+			if($SM.get('thieves')){
+				$SM.set('game.thieves', $SM.get('thieves'));
+				$SM.remove('thieves');
+			}
+			if($SM.get('stolen')){
+				$SM.set('game.stolen', $SM.get('stolen'));
+				$SM.remove('stolen');
+			}
+			if($SM.get('cityCleared')){
+				$SM.set('character.cityCleared', $SM.get('cityCleared'));
+				$SM.remove('cityCleared');
+			}
+			$SM.set('version', 1.3);
+		}
+	},
+
+	/******************************************************************
+	 * Start of specific state functions
+	 ******************************************************************/
+	//PERKS
+	addPerk: function(name) {
+		$SM.set('character.perks["'+name+'"]', true);
+		Notifications.notify(null, Engine.Perks[name].notify);
+	},
+
+	hasPerk: function(name) {
+		return $SM.get('character.perks["'+name+'"]');
+	},
+
+	//INCOME
+	setIncome: function(source, options) {
+		var existing = $SM.get('income["'+source+'"]');
+		if(typeof existing != 'undefined') {
+			options.timeLeft = existing.timeLeft;
+		}
+		$SM.set('income["'+source+'"]', options);
+	},
+
+	getIncome: function(source) {
+		var existing = $SM.get('income["'+source+'"]');
+		if(typeof existing != 'undefined') {
+			return existing;
+		}
+		return {};
+	},
+
+	collectIncome: function() {
+		var changed = false;
+		if(typeof $SM.get('income') != 'undefined' && Engine.activeModule != Space) {
+			for(var source in $SM.get('income')) {
+				var income = $SM.get('income["'+source+'"]');
+				if(typeof income.timeLeft != 'number')
+				{
+					income.timeLeft = 0;
+				}
+				income.timeLeft--;
+
+				if(income.timeLeft <= 0) {
+					Engine.log('collection income from ' + source);
+					if(source == 'thieves') $SM.addStolen(income.stores);
+
+					var cost = income.stores;
+					var ok = true;
+					if (source != 'thieves') {
+						for (var k in cost) {
+							var have = $SM.get('stores["' + k + '"]', true);
+							if (have + cost[k] < 0) {
+								ok = false;
+								break;
+							}
+						}
+					}
+
+					if(ok){
+						$SM.addM('stores', income.stores, true);
+					}
+					changed = true;
+					if(typeof income.delay == 'number') {
+						income.timeLeft = income.delay;
+					}
+				}
+			}
+		}
+		if(changed){
+			$SM.fireUpdate('income', true);
+		}
+		Engine._incomeTimeout = Engine.setTimeout($SM.collectIncome, 1000);
+	},
+
+	//Thieves
+	addStolen: function(stores) {
+		for(var k in stores) {
+			var old = $SM.get('stores["'+k+'"]', true);
+			var short = old + stores[k];
+			//if they would steal more than actually owned
+			if(short < 0){
+				$SM.add('game.stolen["'+k+'"]', (stores[k] * -1) + short);
+			} else {
+				$SM.add('game.stolen["'+k+'"]', stores[k] * -1);
+			}
+		}
+	},
+
+	startThieves: function() {
+		$SM.set('game.thieves', 1);
+		$SM.setIncome('thieves', {
+			delay: 10,
+			stores: {
+				'wood': -10,
+				'fur': -5,
+				'meat': -5
+			}
+		});
+	},
+
+	//Misc
+	num: function(name, craftable) {
```

Comment:
The name of this function is quite vague. It should be changed to reflect its purpose.  Num could reflect anything that is being kept track of or counted in the game. Anyone looking at the code could think it returns the number of attacks from a particular weapon,  the amount of weapons available in the game,  or the amount of objects that the player has. The function name could be changed to num_item or item_amount to show that it's returning the amount of items or objects.